File: backend/agent/obligations/match.go
package obligations

import (
	"regexp"
	"strings"
)

var curlCmdRE = regexp.MustCompile(`(?i)\bcurl(?:\s+-[\w-]+)*\s+https?://[^\s"'` + "`" + `]+`)

func ExtractTaskVerifyCommands(prompt string) []string {
	seen := map[string]bool{}
	var out []string
	add := func(cmd string) {
		cmd = strings.TrimSpace(cmd)
		if cmd == "" || seen[cmd] {
			return
		}
		seen[cmd] = true
		out = append(out, cmd)
	}

	for _, m := range curlCmdRE.FindAllString(prompt, -1) {
		add(m)
	}

	parts := strings.Split(prompt, "`")
	for i := 1; i < len(parts); i += 2 {
		part := strings.TrimSpace(parts[i])
		if LooksLikeVerifyCommand(part) {
			add(part)
		}
	}
	return out
}

func LooksLikeVerifyCommand(command string) bool {
	lower := strings.ToLower(strings.TrimSpace(command))
	if lower == "" {
		return false
	}
	if strings.HasPrefix(lower, "curl ") || strings.Contains(lower, " curl ") {
		return true
	}
	for _, prefix := range []string{
		"go test", "pytest", "npm test", "pnpm test", "yarn test",
		"cargo test", "make test", "vitest", "jest",
	} {
		if strings.HasPrefix(lower, prefix) {
			return true
		}
	}
	return false
}

func IsVerifyCommand(command, verifyCmd string, extra []string) bool {
	if commandMatchesAny(command, verifyCmd, extra) {
		return true
	}
	return LooksLikeVerifyCommand(command)
}

func commandMatchesAny(command, verifyCmd string, extra []string) bool {
	if verifyCmd != "" && commandMatchesPattern(command, verifyCmd) {
		return true
	}
	for _, pattern := range extra {
		if commandMatchesPattern(command, pattern) {
			return true
		}
	}
	return false
}

func commandMatchesPattern(command, pattern string) bool {
	command = strings.TrimSpace(command)
	pattern = strings.TrimSpace(pattern)
	if pattern == "" {
		return false
	}
	if strings.Contains(command, pattern) {
		return true
	}
	if url := extractCurlURL(pattern); url != "" && strings.Contains(command, url) {
		return true
	}
	return false
}

func extractCurlURL(command string) string {
	fields := strings.Fields(command)
	for _, f := range fields {
		if strings.HasPrefix(f, "http://") || strings.HasPrefix(f, "https://") {
			return strings.Trim(f, `"'`)
		}
	}
	return ""
}

File: backend/agent/obligations/match.ts
// PORT: backend/agent/obligations/match.go

const curl_cmd_re = /\bcurl(?:\s+-[\w-]+)*\s+https?:\/\/[^\s"'`]+/gi;

export const extract_task_verify_commands = (prompt: string): string[] => {
  const seen = new Set<string>();
  const out: string[] = [];
  const add = (cmd: string): void => {
    cmd = cmd.trim();
    if (cmd === "" || seen.has(cmd)) {
      return;
    }
    seen.add(cmd);
    out.push(cmd);
  };

  for (const m of prompt.match(curl_cmd_re) ?? []) {
    add(m);
  }

  const parts = prompt.split("`");
  for (let i = 1; i < parts.length; i += 2) {
    const part = parts[i].trim();
    if (looks_like_verify_command(part)) {
      add(part);
    }
  }
  return out;
};

export const looks_like_verify_command = (command: string): boolean => {
  const lower = command.trim().toLowerCase();
  if (lower === "") {
    return false;
  }
  if (lower.startsWith("curl ") || lower.includes(" curl ")) {
    return true;
  }
  for (const prefix of [
    "go test",
    "pytest",
    "npm test",
    "pnpm test",
    "yarn test",
    "cargo test",
    "make test",
    "vitest",
    "jest",
  ]) {
    if (lower.startsWith(prefix)) {
      return true;
    }
  }
  return false;
};

export const is_verify_command = (
  command: string,
  verify_cmd: string,
  extra: string[],
): boolean => {
  if (command_matches_any(command, verify_cmd, extra)) {
    return true;
  }
  return looks_like_verify_command(command);
};

export const command_matches_any = (
  command: string,
  verify_cmd: string,
  extra: string[],
): boolean => {
  if (verify_cmd !== "" && command_matches_pattern(command, verify_cmd)) {
    return true;
  }
  for (const pattern of extra) {
    if (command_matches_pattern(command, pattern)) {
      return true;
    }
  }
  return false;
};

export const command_matches_pattern = (command: string, pattern: string): boolean => {
  command = command.trim();
  pattern = pattern.trim();
  if (pattern === "") {
    return false;
  }
  if (command.includes(pattern)) {
    return true;
  }
  const url = extract_curl_url(pattern);
  if (url !== "" && command.includes(url)) {
    return true;
  }
  return false;
};

export const extract_curl_url = (command: string): string => {
  const fields = command.split(/\s+/);
  for (const f of fields) {
    if (f.startsWith("http://") || f.startsWith("https://")) {
      return f.replace(/["']/g, "");
    }
  }
  return "";
};

/*
PORT STATUS
source path: backend/agent/obligations/match.go
source lines: 97
draft lines: 121
confidence: high
status: phase_b_compile
todos:
  - confirm the RegExp flag (?i) matches Go's case-insensitive matching
notes:
  - No (T, error) returns; plain function port.
*/
